### PR TITLE
Allow covariants of __enter__, __aenter__, and @classmethod

### PR DESCRIPTION
**Pull request by Congee**
The problem we currently have is the return type of classes such as
Client does not allow covariants when subclassing or context manager.
In other words:

```python
class Base:
    def __enter__(self) -> Base:  # XXX
        return self

class Derived(Base):
    ...

with Derived() as derived:
   # The type of derived is Base but not Derived. It is WRONG
    ...
```

There are three approaches to improve type annotations.
1. Just do not type-annotate and let the type checker infer
   `return self`.
2. Use a generic type with a covariant bound
   `_AsyncClient = TypeVar('_AsyncClient', bound=AsyncClient)`
3. Use a generic type `T = TypeVar('T')` or `Self = TypeVar('Self')`

They have pros and cons.
1. It just works and is not friendly to developers as there is no type
   annotation at the first sight. A developer has to reveal its type via
   a type checker. Aslo, documentation tools that rely on type
   annotations lack the type. I haven't found any python docuementation
   tools that rely on type inference to infer `return self`. There are
   some tools simply check annotations.

2. This approach is correct and has a nice covariant bound that adds
   type safety. It is also nice to documentation tools and _somewhat_
   friendly to developers. Type checkers, pyright that I use, always
   shows the the bounded type '_AsyncClient' rather than the subtype.
   Aslo, it requires more key strokes. Not good, not good.

   It is used by `BaseException.with_traceback`
   See https://github.com/python/typeshed/pull/4298/files

3. This approach always type checks, and I believe it _will_ be the
   official solution in the future. Fun fact, Rust has a Self type
   keyword. It is slightly unfriendly to documentation, but is simple to
   implement and easy to understand for developers. Most importantly,
   type checkers love it.

   See https://github.com/python/mypy/issues/1212

But, we can have 2 and 3 combined:

```python
_Base = typing.TypeVar('_Base', bound=Base)

class Base:
   def __enter__(self: _Base) -> _Base:
      return self

class Derive(Base): ...

with Derived() as derived:
   ...  # type of derived is Derived and it's a subtype of Base
```

### CODE REVIEW

**Comment by Congee** [comment]
Hi, this PR is a follow up to #1333 . This PR introduces a typing problem for `StreamContextManager`, it returns a `httpx.Response`. I am not sure how to type it. How would you like it to be typed?

File: httpx/_client.py [hunk 1]
```diff
@@ -1800,7 +1811,7 @@ def __init__(
         self.timeout = timeout
         self.close_client = close_client
 
-    def __enter__(self) -> "Response":
```
Comment (tomchristie) [inline httpx/_client.py:1803]: I don't think there's any reason to change this line.

File: httpx/_client.py [hunk 2]
```diff
@@ -1822,7 +1833,7 @@ def __exit__(
         if self.close_client:
             self.client.close()
 
-    async def __aenter__(self) -> "Response":
```
Comment (tomchristie) [inline httpx/_client.py:1825]: Or this one?

File: httpx/_client.py (hunk 1 above)
Comment (Congee) [reply httpx/_client.py:1803]: Yeah, if nobody is going to subclass StreamContextManager

Comment (Congee) [reply httpx/_client.py:1803]: Will revert this part back

File: httpx/_client.py (hunk 2 above)
Comment (Congee) [reply httpx/_client.py]: reverted

File: httpx/_client.py [hunk 3]
```diff
@@ -493,6 +493,9 @@ def _redirect_stream(
         return request.stream
 
 
+_Client = typing.TypeVar("_Client", bound="Client")
```
Comment (tomchristie) [inline httpx/_client.py:496]: Wondering if we should have a comment here linking to https://www.python.org/dev/peps/pep-0484/#annotating-instance-and-class-methods (?)

Comment (Congee) [reply httpx/_client.py:501]: Did, and added some reasons behind the change here 😉

Comment (johtso) [reply httpx/_client.py:501]: What about just naming it `ClientType`?

Comment (johtso) [reply httpx/_client.py:501]: Or maybe that would imply `Type[Client]`..

Comment (johtso) [reply httpx/_client.py:501]: May just be bikeshedding, if so ignore me. I'm not sure if there is actually a name that neatly encapsulates the concept of an upper bounded type variable. Maybe not worth trying :)

Comment (johtso) [reply httpx/_client.py:501]: Possible colors to paint the shed: `TClient` `AnyClient`

Comment (tomchristie) [reply httpx/_client.py:501]: I guess(?) I'd probably prefer that we follow the PEP and stick with `T`.

Comment (Congee) [reply httpx/_client.py:501]: So, what's the conclusion here? I can make a change

Comment (florimondmanca) [reply httpx/_client.py:501]: I'd also personally go for `T` (or `C`, or whatevs), as `_Client` feels more like a "private base class" than a "generic bounded type variable".

Comment (Congee) [reply httpx/_client.py:501]: Will make a change. Keep in mind that switching to `T = TypeVar('T', covariant=True)` loses the ability to use `TypeVar(.., bound=...)`. After all, `T` is a generic name used everywhere and python's typing system does not support namespace.

Comment (Congee) [reply httpx/_client.py]: changes pushed

Comment (Congee) [reply httpx/_client.py]: Oops, without `bound=...`, mypy complains

Comment (Congee) [reply httpx/_client.py]: As I predicted, lacking of namespace causes trouble. The following works:
```python
T = typing.TypeVar("T", bound="Client")
U = typing.TypeVar("U", bound="AsyncClient")
```
What do you think?

**Review by tomchristie — APPROVED** [review]

**Review by florimondmanca — APPROVED** [review]

**Pull request merged**